lib/notifications/display.dart

```diff
@@ -202,6 +205,61 @@ class NotificationDisplayManager {
     );
   }
 
+  static void _onRemoveFcmMessage(RemoveFcmMessage data) async {
+    assert(debugLog('notif remove zulipMessageIds: ${data.zulipMessageIds}'));
+
+    final groupKey = _groupKey(data);
+
+    var haveRemaining = false;
+    final activeNotifications = await ZulipBinding.instance.androidNotificationHost
+      .getActiveNotifications(desiredExtras: [kExtraZulipMessageId]);
+    for (final statusBarNotification in activeNotifications) {
+      if (statusBarNotification == null) continue; // TODO(pigeon) eliminate this case
+      final notification = statusBarNotification.notification;
+
+      // Sadly we don't get toString on Pigeon data classes: flutter#59027
+      assert(debugLog('  existing notif'
+        ' id: ${statusBarNotification.id}, tag: ${statusBarNotification.tag},'
+        ' notification: (group: ${notification.group}, extras: ${notification.extras}))'));
+
+      // Don't act on notifications that are for other Zulip accounts/identities.
+      if (notification.group != groupKey) continue;
+
+      // Don't act on the summary notification for the group.
+      if (statusBarNotification.tag == groupKey) continue;
+
+      final lastMessageIdStr = notification.extras[kExtraZulipMessageId];
+      if (lastMessageIdStr == null) continue;
+      final lastMessageId = int.parse(lastMessageIdStr, radix: 10);
+      if (data.zulipMessageIds.contains(lastMessageId)) {
+        // The latest Zulip message in this conversation was read.
+        // That's our cue to cancel the notification for the conversation.
+        await ZulipBinding.instance.androidNotificationHost
+          .cancel(tag: statusBarNotification.tag, id: statusBarNotification.id);
+      } else {
```

Comment:
The draft had a `debugLog` line here. Let's keep it; I think it's not any noisier than other debugLog lines I'm glad to have, like on Zulip events.

(And we definitely want it given we have the other debugLog lines above in this method — without this line, those just leave the reader in suspense without knowing what actually happened :slightly_smiling_face:)